iceoryx_utils/include/iceoryx_utils/posix_wrapper/signal_handler.hpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_UTILS_POSIX_WRAPPER_SIGNAL_HANDLER_HPP
+#define IOX_UTILS_POSIX_WRAPPER_SIGNAL_HANDLER_HPP
+
+#include "iceoryx_utils/platform/signal.hpp"
+
+namespace iox
+{
+namespace posix
+{
+using SignalHandlerCallback_t = void (*)(int);
+
+/// @brief Corresponds to the SIG* macros defined in signal.h. The integer values
+///        are equal to the corresponding macro value.
+enum class Signal : int
+{
+    BUS = SIGBUS,
+    INT = SIGINT,
+    TERM = SIGTERM
+    /// @attention never add SIGKILL or SIGSTOP into this list, they cannot be caught
+    ///            and sigaction returns the errno EINVAL
+};
+
+/// @attention NEVER USE THIS CLASS AS A MEMBER VARIABLE! A class which should be used
+///            only in method/function scopes.
+/// @brief The SignalGuard is a class returned by registerSignalHandler. When it goes
+///         out of scope it restores the previous signal action. Typical use case:
+///         One would like to override the signal action in main() or some C posix
+///         makes it necessary to override the standard signal action before and
+///         after the call.
+/// @code
+///    {
+///      auto sigHandler = registerSignalHandler(Signal::BUS, printErrorMessage);
+///      my_c_call_which_can_cause_SIGBUS();
+///    }
+///    // here we are out of scope and the signal action for Signal::BUS is restored
+/// @endcode
+class SignalGuard
+{
+  public:
+    SignalGuard() noexcept = default;
+    SignalGuard(const Signal signal, const struct sigaction& previousAction) noexcept;
```

Comment:
Please add documentation.

Should this be public? Maybe it's less error prone if it can only be used by `registerSignalHandler´